### PR TITLE
Staking refactor. Change *_or() to *_or_else() 

### PR DESCRIPTION
**Pull request by Mr-Leshiy**
According to this articles https://newbedev.com/why-should-i-prefer-option-ok-or-else-instead-of-option-ok-or , https://stackoverflow.com/questions/45547293/why-should-i-prefer-optionok-or-else-instead-of-optionok-or it is preferable to use *_or_else instead of *_or to avoid a redundant memory allocation if it is not needed.

### CODE REVIEW

File: frame/staking/src/benchmarking.rs [hunk 1]
```diff
@@ -134,25 +134,25 @@ benchmarks! {
 	bond_extra {
 		let (stash, controller) = create_stash_controller::<T>(USER_SEED, 100, Default::default())?;
 		let max_additional = T::Currency::minimum_balance() * 10u32.into();
-		let ledger = Ledger::<T>::get(&controller).ok_or("ledger not created before")?;
```
Comment (bkchr) [inline frame/staking/src/benchmarking.rs:137]: None of the changes in this file makes any sense. This is a static string and nothing will be allocated here.

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -2315,8 +2315,8 @@ impl<T: Config> Pallet<T> {
 
 	fn do_payout_stakers(validator_stash: T::AccountId, era: EraIndex) -> DispatchResultWithPostInfo {
 		// Validate input data
-		let current_era = CurrentEra::<T>::get().ok_or(
-			Error::<T>::InvalidEraToReward.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
+		let current_era = CurrentEra::<T>::get().ok_or_else(
+			|| Error::<T>::InvalidEraToReward.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
```
Comment (bkchr) [inline frame/staking/src/lib.rs:2319]: This.

File: frame/staking/src/lib.rs [hunk 3]
```diff
@@ -2332,8 +2332,8 @@ impl<T: Config> Pallet<T> {
 					.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
 			)?;
 
-		let controller = Self::bonded(&validator_stash).ok_or(
-			Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
+		let controller = Self::bonded(&validator_stash).ok_or_else(
+			|| Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
```
Comment (bkchr) [inline frame/staking/src/lib.rs:2336]: And this are probably the only relevant ones.

**Comment by Mr-Leshiy** [comment]
@bkchr , thank you for your response.
What do you think about this one
```
let mut ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
let mut ledger = Self::ledger(&controller).ok_or_else(|| Error::<T>::NotController)?;
```
I assume that these errors are also statically initialized

**Comment by bkchr** [comment]
> @bkchr , thank you for your response.
> What do you think about this one
> 
> ```
> let mut ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
> let mut ledger = Self::ledger(&controller).ok_or_else(|| Error::<T>::NotController)?;
> ```
> 
> I assume that these errors are also statically initialized

Yes.

File: frame/staking/src/benchmarking.rs (hunk 1 above)
Comment (Mr-Leshiy) [reply frame/staking/src/benchmarking.rs]: Fixed.

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -3110,7 +3110,7 @@ impl<T: Config> frame_election_provider_support::ElectionDataProvider<T::Account
 		targets.into_iter().for_each(|v| {
 			let stake: BalanceOf<T> = target_stake
 				.and_then(|w| <BalanceOf<T>>::try_from(w).ok())
-				.unwrap_or(MinNominatorBond::<T>::get() * 100u32.into());
+				.unwrap_or_else(|| MinNominatorBond::<T>::get() * 100u32.into());
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:3113]: only this one might be sensible

**Review by kianenigma — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@Mr-Leshiy please merge master

**Comment by bkchr** [comment]
bot merge

**Pull request merged**